### PR TITLE
Address problems with Spawn module

### PR DESCRIPTION
**Pull request by mppf**
In preparing an example for the SC tutorial, I tried
using the Spawn module where the Chapel program has
a task providing input to a sub-process and another
task consuming the output. This did not work.

It now works. The new test cat-test.chpl shows the pattern.

      var sub = spawn(["cat"], stdin=PIPE, stdout=PIPE);
      cobegin {
	{
	  for x in test {
	    sub.stdin.writeln(x);
	  }
	  sub.stdin.close();
	}

	{
	  var line:string;
	  var i = 1;
	  while sub.stdout.readln(line) {
	    if verbose then writeln("Got line ", line);
	    assert(line == test[i]);
	    i += 1;
	  }
	}
      }
      sub.wait();

In order to get this to work, I needed to make 2 changes:

1) Previously, in spawn(), we would *mark* the sub.stdin channel
   if there was also a PIPE for stout or stderr. That way,
   we could write to the input channel without worrying that
   the subprocess will block waiting for us to read from it.

   However, in the pattern above with cobegin, we need
   the input channel not to be 'mark'd in some special
   situation. (if it is... the example above won't ever
   write to the pipe).

   Therefore, we add a new constant in the Spawn module,
   BUFFERED_PIPE, which can be used for stdin to trigger
   the buffering behavior.

2) Previously, the subprocess record had fields for both
   the *file* and the *channel* for each of stdin,stdout,stderr.
   That caused problems with the pattern above because the
   call to `sub.stdin.close()` would not actually close
   the file (since it's reference count reflects that the
   subprocess record has a file field referring to it). That
   meant that the program above would hang, because `cat'
   won't complete until its input is closed.

   So, I updated the subprocess record and associated
   code to only track the channel (and to get the file
   from the channel when it is needed).

While there, I noticed that closing a channel that has sole
ownership over a file will close the file, but doesn't
directly return an error if there was one. Fixed that.

- test/io/ferguson and test/modules/standard/Spawn on Mac OS X
- passed full local testing on linux
- Reviewed by @lydia-duncan - thanks!

### CODE REVIEW

**Comment by mppf** [comment]
@lydia-duncan - are you interested in reviewing this one?

File: modules/standard/Spawn.chpl [hunk 1]
```diff
@@ -80,13 +79,47 @@ back its input.
...
+  var input = ["a", "b", "c"];
+
+  var sub = spawn(["cat"], stdin=PIPE, stdout=PIPE);
+  cobegin {
+    {
+      // one task writes data to the subprocess
+      for x in input {
+        sub.stdin.writeln(x);
+      }
+      // this close is important so that the kkkkkkk
```
Comment (lydia-duncan) [inline modules/standard/Spawn.chpl]: Did you mean to replace "kkkkkkk" with something else?

File: modules/standard/Spawn.chpl [hunk 2]
```diff
@@ -443,9 +480,12 @@ module Spawn {
...
         }
 
-    if stdin == QIO_FD_PIPE then stdin_pipe = true;
-    if stdout == QIO_FD_PIPE then stdout_pipe = true;
-    if stderr == QIO_FD_PIPE then stderr_pipe = true;
+    if stdin == QIO_FD_PIPE || stdin == QIO_FD_BUFFERED_PIPE then
+      stdin_pipe = true;
+    if stdout == QIO_FD_PIPE || stdout == QIO_FD_BUFFERED_PIPE then
+      stdout_pipe = true;
+    if stderr == QIO_FD_PIPE || stderr == QIO_FD_BUFFERED_PIPE then
```
Comment (lydia-duncan) [inline modules/standard/Spawn.chpl]: Might be worth making this a function that checks if the channel provided is a pipe of any sort, but it's not a big deal

**Comment by lydia-duncan** [comment]
I think this is good to go now!

**Pull request merged**